### ISSUE TITLE
Symfony 6 IsGranted at the controller level regression

### ISSUE THREAD

**Issue by justmoen** [symfony/symfony#51361]
### Description

The following annotation applied at the top of the controller granted access to all controller actions, prior to the deprecation of @Security.

```
/**
 * @Security("is_granted('view', 'PermissionModuleName'))
 */
class SomeController {}
```

To allow the following, equivalent functionality, 

```
#[IsGranted('view', 'PermissionModuleName')]
class SomeController {}
```

change IsGrantedAttributeListener->getIsGrantedSubject() method to 

`return $arguments[$subjectRef] ?? $subjectRef`

**Comment by derrabus** [comment]
I don't see a regression here.

**Comment by justmoen** [comment]
#[IsGranted('view', 'PermissionModuleName')] does not function at the controller level the same way that @Security(is_granted()) did.  Now, the attribute must be added to every separate controller action when I am expecting all controller actions to have the same permission restriction.

The only other way to achieve this is by custom Expression

`#[IsGranted(new Expression("is_granted('view', 'PermissionModuleName')))]`

or overriding the IsGrantedAttributeListener with the change mentioned in OP

**Comment by derrabus** [comment]
> The only other way to achieve this is by custom Expression
> 
> `#[IsGranted(new Expression("is_granted('view', 'PermissionModuleName')))]`

There you go. That's the solution.

**Comment by aless673** [comment]
> > The only other way to achieve this is by custom Expression
> > `#[IsGranted(new Expression("is_granted('view', 'PermissionModuleName')))]`
> 
> There you go. That's the solution.

That's a critical regression and the issue should not be marked as closed.

The documentation clearly show the possibility to use IsGranted at the controller level : https://symfony.com/doc/current/security.html#security-securing-controller-attributes

It also means that any project upgrading to Symfony 6.x will have breaches in their security as the attribute IsGranted is not working as it was in previous versions. There are no warnings, no errors. That's sounds terrible.

**Comment by nicolas-grekas** [comment]
Someone needs to send a PR if any change is suggested.

**Comment by derrabus** [comment]
> That's a critical regression and the issue should not be marked as closed.

Maybe we're fighting over terminology here, but… This would be a regression if without any code modification, an app would stop working after a Symfony upgrade.

In the code we see in the bug description, the author migrated from an annotation of an external bundle (FrameworkExtraBundle) to an attribute provided by Symfony core. Those are two different features of two different packages.

> It also means that any project upgrading to Symfony 6.x will have breaches in their security as the attribute IsGranted is not working as it was in previous versions. There are no warnings, no errors. That's sounds terrible.

No. The `IsGranted` attribute of `symfony/security-core` did not exist in Symfony 5 and FrameworkExtraBundle still works with Symfony 6. You need to actively change your code to run into the problems described by the author.

**Comment by aless673** [comment]
> > That's a critical regression and the issue should not be marked as closed.
> 
> Maybe we're fighting over terminology here, but… This would be a regression if without any code modification, an app would stop working after a Symfony upgrade.
> 
> In the code we see in the bug description, the author migrated from an annotation of an external bundle (FrameworkExtraBundle) to an attribute provided by Symfony core. Those are two different features of two different packages.
> 
> > It also means that any project upgrading to Symfony 6.x will have breaches in their security as the attribute IsGranted is not working as it was in previous versions. There are no warnings, no errors. That's sounds terrible.
> 
> No. The `IsGranted` attribute of `symfony/security-core` did not exist in Symfony 5 and FrameworkExtraBundle still works with Symfony 6. You need to actively change your code to run into the problems described by the author.

- Symfony doc 5.x explains to use Sensio's "IsGranted" annotation in 5.x (https://symfony.com/doc/5.x/security.html#securing-controllers-and-other-code)
- Symfony doc 6.x explains to use their attribute "IsGranted" (for the same purpose, at the same doc section)
- SensioFrameworkExtraBundle package is archived and invite developers to use Symfony core attributes instead.
- Rector set "symfony62" includes a renaming of the import from Sensio to Symfony
- Symfony blog talks about migrating from Sensio to Symfony : https://symfony.com/blog/new-in-symfony-6-2-built-in-cache-security-template-and-doctrine-attributes

**It is a mistake to not consider this issue as critical.**

**Comment by derrabus** [comment]
> * Symfony doc 5.x explains to use Sensio's "IsGranted" annotation in 5.x (https://symfony.com/doc/5.x/security.html#securing-controllers-and-other-code)

Which the author didn't use. They used the `@Security` annotation and migrated to the `#[IsGranted]` attribute, which didn't work the way they expected.

> * SensioFrameworkExtraBundle package is archived and invite developers to use Symfony core attributes instead.

Yes, that is the idea. However, it is not required for a Symfony 6 upgrade. If you just do the Symfony upgrade, your annotations should continue to work as they did before.

Once the upgrade is done, you can replace the FEB annotations with Symfony's attributes, one by one. In case of `@Security` annotations with expressions, this involves a little extra work as explained above.

> * Rector set "symfony62" includes a renaming of the import from Sensio to Symfony

If rector really renames `@Security` to `#[IsGranted]` without further changes (like wrapping expressions into an `Expression` object, please send a bug report to Rector.

> **It is a mistake to not consider this issue as critical.**

Maybe it's better if you open a new issue and explain _your_ problem. Apparently, your problem is different from the issue's author's.

**Comment by maximehardy** [comment]
@derrabus what you are suggesting is that everyone other than you is wrong and should change its way of doing it.
Let's be realistic, if the ecosystem has anchored this behaviour, now you can't change it by snapping your fingers.
I do agree with @aless673, this issue is **critical**. We are talking about changing a **security** behaviour, not the colour of a button!

**Comment by derrabus** [comment]
> @derrabus what you are suggesting is that everyone other than you is wrong and should change its way of doing it.

This is not what I'm suggesting and I won't have this conversation with you. We can have a debate, but attacking people personally won't get you anything.

> We are talking about changing a **security** behaviour, not the colour of a button!

No behavior has been changed.

**Comment by maximehardy** [comment]
You are not having a debate, you simply refuted the facts presented by others with no opening. This is no personal attack, this is an observation.

Now about the fact that behaviour changed, see #54542, it indeed changed in an unexpected way. And it seems clear this is a bug.

**Comment by derrabus** [comment]
> Now about the fact that behaviour changed, see #54542, it indeed changed in an unexpected way. And it seems clear this is a bug.

Sure, but that is a different issue, isn't it.
